File: Ruby/problem060.rb
# -*- coding: utf-8 -*-
class Problem060
  class << self
    def can_partition?(arr)
      if arr.sum.even?
        partition_sum = arr.sum / 2
        buf = 0
        sorted_arr = arr.sort

        while buf < partition_sum
          buf += sorted_arr.pop
        end

        buf == partition_sum
      else
        false
      end
    end

    def find_partition_one_elem(arr)
      # Assuming the array can be partitioned, return the element which is
      # equal to arr.sum /2
      arr.select { |x| x == arr.sum / 2 }
    end

    def find_partition_two_elems(arr)
      partition_sum = arr.sum / 2
      arr.select { |n| arr.include?(partition_sum - n) }
    end
  end
end

File: JavaScript/problem060.js
const canPartition = (arr) => {
  if (arr.reduce((x, y) => x + y) % 2 === 0) {
    let partitionSum = arr.reduce((x, y) => x + y) / 2
    let out = []

    let buf = partitionSum
    let numberSpace = [...new Set(arr)]

    while (buf > 0) {
      // Get the max element of the unique array.
      let current = Math.max(...numberSpace)
      out.push(current)
      numberSpace.splice(numberSpace.indexOf(current), 1)

      if (arr.includes(buf - current)) {
        return [...out, buf - current].reduce((x, y) => x + y) == partitionSum
      }

      buf -= current
    }

    return false
  }

  return false
}

module.exports = {
  canPartition
};
